### ISSUE TITLE
Refactor proposals for Property value preference linters

### ISSUE THREAD

**Issue by srawlins** [brigade/scss-lint#769]
I'm looking at proposed linters like #767 ("border value order"), #741 ("font-weight notation"), #733 ("disallow visibility:visible") and existing linters like BorderZero, Shorthand, TransitionAll, and I see a few problems:

1. Implementing each of these as their own linter, and enabling them all requires **visiting the entire tree** of an SCSS file over and over again, just to call the same visit methods over and over.
2.  Implementing each of these is tedious.
3.  When scss-lint hits 1.0, and all linters are disabled by default, reviewing all of them as a new user would be crazy tedious. "TransitionAll, BorderZero... whatever, I just want to linty my SCSS!"

I can also see a possibility for how these could all be configured in a single PropertyValue linter:

```yaml
linters:
  PropertyValue:
    enabled: true
    rules:
      - property_pattern: "border(-top|-right|-bottom|-left)?"
        banned_pattern: "none"
        message: "{property_pattern}: 0 is preferred over none"
      - property_pattern: "font-weight"
        allowed_pattern: "[1-8]00"
        message: "font-weight: prefer numerical weight"
      - property_pattern: "transition"
        banned_pattern: "all"
        message: "transition: all is non-performant"
      - property_pattern: "border(-top|-right|-bottom|-left)?"
        allowed_pattern: "<length> <border-style> <color>"
```

At this point, looking at this proposal, it just reeks of overengineering, and it's probably a _less_ friendly interface for the user, albeit pretty powerful. It doesn't help with super generic restrictions, like ColorKeyword, ColorVariable, HexLength,LeadingZero, TrailingZero.

It also puts messaging in the user's hands, which would be a first. And it would be hard for scss-lint to provide "default settings"; how would you override just one?

Maybe this proposal could be limited to straight up simple bans. The config becomes simpler:

```yaml
linters:
  PropertyValue:
    enabled: true
    rules:
      - property_pattern: "border(-top|-right|-bottom|-left)?"
        banned_pattern: none
      - property_pattern: font-weight
        allowed_pattern: "[1-8]00"
      - property_pattern: transition
        banned_pattern: all
      - property_pattern: visibility
        banned_pattern: visible
```

And then the messaging is derived from the `allowed_pattern` or `banned_pattern`. This is nicer, more concise. The inheritance problem persists...

I think the original problem is not a small one though: these "fringe" linters provide less value, and are probably not important to most projects, but add to the growing list of linters to enable or not enable.

With the availability of "Custom Linters", it'd be great if people could just write their own linters. You think, "how hard can it be to write a 'ban visibility: visible' linter? 5 lines?" Then you look at [transition_all.rb](https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/transition_all.rb) to see that it took 30 lines to write that linter properly.

You can imagine this as a simpler linter if inherited from an abstract PropertyValue linter:

```ruby
module SCSSLint
  class Linter::TransitionAll < Linter::PropertyValue
    def properties
      %w[transition, transition-property].freeze
    end

    def check_it(node, property, value)
      return unless offset = value =~ /\ball\b/

      pos = node.value_source_range.start_pos.after(value[0, offset])

      add_lint(Location.new(pos.line, pos.offset, 3),
               "#{property} should contain explicit properties instead of using the keyword all")
    end
  end
end
```

That's a lot to chew on. What do y'all think?

**Comment by connorshea** [comment]
This is awesome, I don't know enough about the internals of the gem to comment on how reasonable it'd be to implement, but this is awesome.

**Comment by sds** [comment]
Thanks for taking the time to put together some of your ideas, @srawlins.

I agree it's not ideal from a performance perspective that we're visiting the same nodes over and over again. I also agree it's not great to have very similar implementations and essentially duplicate tests for multiple linters. 

For the performance perspective, I think #576 would "cover up" most of these concerns. However that doesn't address all the almost-duplicated code and tests as well as the frustration you mention developers would have with enabling all the linters they want.

Regarding the format of the configuration: if you key it based on a rule "name" you can override configuration much more easily (modifying default rules or adding your own–removing would require a separate `enabled` option):

```
linters:
  PropertyValue:
    enabled: true
    rules:
      border-none:
        property_pattern: "border(-top|-right|-bottom|-left)?"
        banned_pattern: none
      font-weight:
        property_pattern: font-weight
        allowed_pattern: "[1-8]00"
      ...
```

One thing to keep in mind is that moving this entire set of checks into a single linter limits your ability to change which directories are included/excluded on a per-check basis, since `include`/`exclude` apply to the entire linter. I'm ok with this, but there are always developers who want that fine-grained level of control.

Regarding your idea to provide an abstract hook implementation: I also would be open to this. We'll have to figure out how we exclude these from being run (i.e. explicitly marking them as abstract in some way–maybe storing in a special directory), and how we document them so they are easily discoverable. In general, making it easier for developers to write their own hooks sounds like a good thing.
